### ISSUE TITLE
Gemalto IDPrime 840 v3 unable to do TLS identification with AES 256 GCM SHA 384 server

### ISSUE THREAD

**Issue by kirichkov** [OpenSC/OpenSC#2309]
### Problem Description

The new implementation for IDPrime v3 is having trouble with TLS authentication. Firefox is reporting the server is using TLS 1.2 - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, 256 bit keys).

The error displayed by Firefox is SEC_ERROR_PKCS11_GENERAL_ERROR

TLS authentication works with Gemalto's proprietary driver, and authentication works with OpenSC with (different server) using TLS 1.2 (TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, 128 bit keys)

OS is macOS 11.2.3

### Logs

[Gist with debug = 9](https://gist.github.com/kirichkov/91486113ca3a346b8861401b39d2ccb4)

**Comment by Jakuje** [comment]
The [SW 0x6985](https://gist.github.com/kirichkov/91486113ca3a346b8861401b39d2ccb4#file-opensc-debug-2021-04-25-tr-txt-L4010) is CONDITION_NOT_SATISFIED. There is probably something wrong with the security environment. Can you collect the APDU trace from the proprietary driver?

**Comment by kirichkov** [comment]
Here's the [log of the APDU from the proprietary driver](https://gist.github.com/kirichkov/a5b928c33afb8fae42b1faaff7ccac5a) created on Mac OS.

While trying to get the APDU log, I found out that the proprietary driver the card's seller ships for Linux also couldn't authenticate. I've uploaded the [linux log](https://gist.github.com/kirichkov/7294788622dba5ac702551237ed30edf) for reference too.

**Comment by Jakuje** [comment]
the proprietary driver:

    00 22 41 B6 06 80 01 02 84 01 F8

the opensc:
 
    00 22 41 b6 06 80 01 42 84 01 f8

The difference is in 02 and 42 byte. The Firefox is trying to use CKM_RSA_PKCS mechanism:

    P:1169; T:0x6130987008 16:16:27.882 [onepin-opensc-pkcs11] mechanism.c:256:sc_pkcs11_sign_init: mechanism 0x1, key-type 0x0

so it looks like it is trying different RSA mechanism with the proprietary driver.

Can you try `pkcs11-tool --test --pin YOUR_PIN` to see if some other mechanisms work?

I think the detection of the mechanisms here is broken anyway and we will need some love:

https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-idprime.c#L653-L661

Would a simple hack like this make the code working for you?
```
diff --git a/src/libopensc/card-idprime.c b/src/libopensc/card-idprime.c
index ba9bb8bc..9a21edce 100644
--- a/src/libopensc/card-idprime.c
+++ b/src/libopensc/card-idprime.c
@@ -650,15 +650,7 @@ idprime_set_security_env(struct sc_card *card,
 				new_env.algorithm_ref = 0x65;
 			}
 		} else { /* RSA-PKCS */
-			if (env->algorithm_flags & SC_ALGORITHM_MGF1_SHA256) {
-				new_env.algorithm_ref = 0x42;
-			} else if (env->algorithm_flags & SC_ALGORITHM_MGF1_SHA384) {
-				new_env.algorithm_ref = 0x52;
-			} else if (env->algorithm_flags & SC_ALGORITHM_MGF1_SHA512) {
-				new_env.algorithm_ref = 0x62;
-			} else { /* RSA-PKCS without hashing */
-				new_env.algorithm_ref = 0x02;
-			}
+			new_env.algorithm_ref = 0x42;
 		}
 		break;
 	default:
```

**Comment by kirichkov** [comment]
Output from `pkcs11-tool --test --pin YOUR_PIN`

```
Using slot 0 with a present token (0x0)
C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  ERR: C_GenerateRandom failed: CKR_DATA_LEN_RANGE (0x21)
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
  RIPEMD160: OK
Signatures (currently only for RSA)
  testing key 0 (Private key 1) 
  all 4 signature functions seem to work
  testing signature mechanisms:
    RSA-PKCS: OK
    SHA256-RSA-PKCS: ERR: verification failed
  testing key 1 (Private key 2) with 1 mechanism
    RSA-PKCS: OK
Verify (currently only for RSA)
  testing key 0 (Private key 1)
    RSA-PKCS: OK
  testing key 1 (Private key 2) with 1 mechanism
    RSA-PKCS: OK
Decryption (currently only for RSA)
  testing key 0 (Private key 1)
    RSA-PKCS: OK
    RSA-PKCS-OAEP: mgf not set, defaulting to MGF1-SHA256
OAEP parameters: hashAlg=SHA256, mgf=MGF1-SHA256, source_type=0, source_ptr=0x0, source_len=0
OK
  testing key 1 (Private key 2)
    RSA-PKCS: OK
    RSA-PKCS-OAEP: mgf not set, defaulting to MGF1-SHA256
OAEP parameters: hashAlg=SHA256, mgf=MGF1-SHA256, source_type=0, source_ptr=0x0, source_len=0
OK
2 errors
```

**Comment by kirichkov** [comment]
> the proprietary driver:
> 00 22 41 B6 06 80 01 02 84 01 F8

Just to confirm that you switched them around. The log above (0x02) is from OpenSC. The one below (0x42) is the proprietary driver. Judging by the patch, I think it was only a wording mismatch.

> the opensc:
> 00 22 41 b6 06 80 01 42 84 01 f8

I tried the patch, now the AES 128 bit also fails :) The bit sent is now 0x00 instead of 0x02 or 0x42, so something is changing afterwards. Last digit is 0x32 instead of F8, but maybe that's a checksum?

**Comment by Jakuje** [comment]
can you check with #2342. We merged recently one fix for the on-card hashing and with the above PR, you should be able to get a bit further (hopefully).

**Comment by kirichkov** [comment]
I've compiled the branch - neither AES128 or AES256 authentication works. [Here's the log](https://gist.github.com/kirichkov/ae3edfb9a6c053bd0f443ea3a57fadcd)

**Comment by Jakuje** [comment]
Hmm ... this fixed the tests for me. Not sure why you get the 00 in the algorithm reference though ...

**Comment by Jakuje** [comment]
> P:10864; T:0x123145417560064 20:32:27.446 [onepin-opensc-pkcs11] framework-pkcs15.c:4000:pkcs15_prkey_sign: Initiating signing operation, mechanism 0xd.

ok, this is `CKM_RSA_PKCS_PSS` so there might be some catch. 

> P:10864; T:0x123145417560064 20:32:27.447 [onepin-opensc-pkcs11] framework-pkcs15.c:4133:pkcs15_prkey_sign: Selected flags 200110. Now computing signature for 32 bytes. 256 bytes reserved.

The flags here mean SC_ALGORITHM_RSA_PAD_PSS | SC_ALGORITHM_RSA_HASH_NONE | SC_ALGORITHM_MGF1_SHA256 (would love to have them translated in the log directly). The 32 bytes is indeed SHA256 hash value.

> P:10864; T:0x123145417560064 20:32:27.447 [onepin-opensc-pkcs11] pkcs15-sec.c:613:sc_pkcs15_compute_signature: supported algorithm flags 0xE0E032, private key usage 0x20E

The private key usage has decrypt, sign, sign_recover, SC_PKCS15_PRKEY_USAGE_NONREPUDIATION, which is fine.

> P:10864; T:0x123145417560064 20:32:27.447 [onepin-opensc-pkcs11] pkcs15-sec.c:671:sc_pkcs15_compute_signature: DEE flags:0x00200110 alg_info->flags:0x00e0e032 pad:0x00000000 sec:0x00000010

We have no padding flags, which is fine. The sec flags is SC_ALGORITHM_RSA_PAD_PSS only, which looks wrong. The MGF1 flags should not have been lost, I think.

Your line numbers are a bit off from the current master. Do you have the commit 44d429c3 in your tree? If not, try with that commit too.

**Comment by kirichkov** [comment]
IT WORKS! I'm very sorry about this! Once I cloned from scratch and recompiled it is now able to do successful TLS authentication with both sites! I thought `make clean` should remove all stale artifacts, but I guess it didn't.

**Comment by Jakuje** [comment]
Thanks for testing again. The `make clean` should do it but there are sometimes weird anomalies. I guess we can close this issue now.

**Comment by Jakuje** [comment]
@kirichkov Do you still have IDPrime card? We are reworking some of the code to support more IDPrime card versions in #2666 and could use some testing and feedback, if you still have the card and you are willing to test the changes.

**Comment by kirichkov** [comment]
I do! Unfortunately I have only ARM-based Mac nowadays, so hopefully that won't be a problem to compile from source.

**Comment by dengert** [comment]
Do you have a Raspberry Pi? OpenSC can be built on it.

**Comment by kirichkov** [comment]
I have a Linux box (amd64) and Raspberry Pi, but my main usage is on a Mac, so I preferred to test on a Mac. I'll test it on Linux, probably around the end of the week, and let you know.

**Comment by kirichkov** [comment]
I compiled and tested the branch - I did a user-level test - file signing and website authentication via Firefox and Chromium. File signing works, authentication works in Firefox. I didn't have any success with authentication in Chromium, but I never used it/set it up in the first place, so I don't think this is in any way related to OpenSC. Most likely adding the opensc library to Chromium wasn't correct!

If you push a nightly with macOS arm64/amd64 support let me know and I'll test it as well!
